[feature][need-docs] Current feature/current value form context expressions

elpaso opened the pull request:
## Form/Attribute table context expressions

This PR implements the QEP 20 https://github.com/qgis/QGIS-Enhancement-Proposals/issues/116 (Allow Drill-down (cascade) search in Value Relation Widgets).

A demo  and a detailed explanation can be found here:  https://north-road.com/drill-down-cascading-forms/

### New functions and variables:

This implementation add two new expression func/vars, available in the attribute table and in the feature forms and that can be used in the Value Relation Widget:


```
current_value('FIELD_NAME')
```
Returns the current form/table-row value of the feature currently being edited

```
@current_geometry
```
Returns the current geometry value of the feature currently being edited

## Tests

Tests have been added for the widget behaviors in a C++ test.
A python test contains tests for the logic which handles the dynamic cache by deciding if the expression contains dynamic values (i.e. if it contains any of the form context expressions and if they are valid) and if they can be used, in case the expression does not require the form context the whole dynamic cache is skipped and the system behave exactly as it behaved before this PR.

elpaso committed:
[feature] Form context expressions in value relation widget

The value relation widget filter expression can now use two
new functions/variables that have access to the current
values and geometry of the form being edited.

This allows for dynamic filtering (drill-down) as explained
in the crowdfunding page:
https://north-road.com/drill-down-cascading-forms/

The new functions/variables are:

Function:
get_current_form_field_value( 'FIELD_NAME' )

Variable:
@current_form_geometry

elpaso committed:
Rename functions from get_current_form_field_value to current_value

and current_geometry

elpaso committed:
Some more test cases for value relation widget

m-kuhn commented on src/gui/editorwidgets/core/qgseditorwidgetwrapper.h:
```diff
@@ -277,6 +277,18 @@ class GUI_EXPORT QgsEditorWidgetWrapper : public QgsWidgetWrapper
+    /**
+     * mFeature the current feature
+     */
+    QgsFeature mFeature;
```
That means the attributes of the feature can be changed without appropriate notifications being sent? I'd prefer to avoid having this directly writable to subclasses and make (protected?) writers instead.
The adding docs is much appreciated, but should include information about which state this feature has (i.e. the feature which was set originally or the current state of the feature on the form.

m-kuhn commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -85,6 +87,12 @@ QVariant QgsValueRelationWidgetWrapper::value() const
 
 QWidget *QgsValueRelationWidgetWrapper::createWidget( QWidget *parent )
 {
+  QgsAttributeForm *form = dynamic_cast<QgsAttributeForm *>( parent );
```
qobject_cast?

m-kuhn commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -184,6 +149,11 @@ void QgsValueRelationWidgetWrapper::setValue( const QVariant &value )
+
+    // This block is needed because item->setCheckState triggers dataChanged gets back to value()
+    // and iterate over all items again! This can be extremely slow on large items sets.
+    mTableWidget->blockSignals( true );
```
Would it be possible to use a variable to guard from this? I'm always afraid of unintended side-effects of blocking all signals of an object.

elpaso replied:
I'm not sure I get what you mean, the logic here is to block signals during the loop and fire the signal only once for the last item.
My test case was probably a worst case scenario where the referenced layer had hundreds of items, this means that in case of AllowMulti we had to loop through hundred of items and set the checkbox for each of them and fire the signal (which is not just one but a chain of signals) for each of them, this proved to be extremely slow, that's why I decided to block the signals in the loop and fire just once for the last item, this solved the performance issue but retains the original behavior because the signal is still fired (just once instead of hundred of times).

m-kuhn replied:
The risk with this approach is that not only the problematic signal is blocked (which you take care of emitting in the end) but every signal. I don't know which other ones it would be here at the moment and no one knows which ones are introduced in the future. That's why I think it's often a good idea to set a member variable (`bool mIsUpdatingWidget`, better name welcome) and check in the slot if it is set and exit. But this only works if it's one of our slots and not a Qt slot.

m-kuhn replied:
No, I guess this falls into the category where it doesn't work.

m-kuhn commented on src/gui/qgsattributeform.h:
```diff
@@ -353,6 +354,8 @@ class GUI_EXPORT QgsAttributeForm : public QWidget
 
     QString createFilterExpression() const;
 
+    //void updateWidgetFeature( QgsEditorWidgetWrapper *w );
```
Leftover?

m-kuhn commented on src/gui/qgsattributeeditorcontext.h:
```diff
@@ -189,6 +190,21 @@ class GUI_EXPORT QgsAttributeEditorContext
+     * \see setFormFeature()
+     * \since QGIS 3.2
+     */
+    const QgsFeature formFeature() const { return mFormFeature; }
```
`QgsFeature formFeature() const`

m-kuhn commented:
Good job @elpaso , nice to see this coming to core :tada: 

What's the reason to use an accessor function for the attribute and a variable for the geometry? Wouldn't it be more consistent to use either one or the other?

elpaso commented:
@m-kuhn thanks for reviewing!
Honestly I though quite a lot about the function/variable and I've not a strong opinion, IIRC I decided for a function (for `current_value`) because it takes an argument and because it looks like the `attribute` function, OTOH `@current_geometry` reminds me of `@geometry`

elpaso commented on src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp:
```diff
@@ -21,6 +21,8 @@
 
 #include <QSettings>
 
+QString QgsValueRelationFieldFormatter::FORM_SCOPE_FUNCTIONS_RE = QStringLiteral( "%1\\s*\\(\\s*'([^']+)'\\s*\\)" );
```
What is the AST?

m-kuhn replied:
The *Abstract Syntax Tree*, i.e. visiting the nodes instead of parsing the plain expression.

elpaso replied:
> Can the AST of the expression nodes be used instead of a regex here?

Yes, I think it can be used (I need to learn how but this is another story).

> if I understand it correctly, this won't allow using layer variables or expression results?
No: the regexp is a template and gets filled with the function names that require a form scope to answer to a simple question: does this expression contain any function that requires a form scope?

I don't see the advantages in using AST over a simple regexp but maybe you can show me the light and give me a little tip about how to do this without writing thousands of lines of code. :smiley: 

I didn't find a tree traversal API in `QgsExpressionNode`: do I really need to try to cast every node starting from the root and branch for every node type? Is there a way to retrieve the list of all **functions** in the tree as there is for **variables**?

m-kuhn replied:
Regex for parsing is not very tolerant `current_value( @main_field )` will probably not work. Plus one needs to take care of any current and upcoming syntax details (inline comments, escape sequences, quoting, ... ) to make it failproof.

You could probably add a method similar to `referencedColumns` and `referencedVariables` to get all function nodes (or even a template function to generically get all nodes by class type, similar to `QList<T> QObject::findChildren()` if you want to get crazy ;) ). Or as you say have some fun with traversing from outside and have some casting fun.

You'll then probably want to use `eval()` on the child param node to get the field name.

nyalldawson replied:
+1 to adding a referencedFunctions method - should be just copy and paste from the existing methods

elpaso replied:
@m-kuhn this is done in my last commit, but it does not solve the problem: once I get the names of the `referencedFunctions` how do I get the (evaluated) param values?

Do I also need a `QList<QgsExpressionFunction> referencedFunctionsObjects() const;` ?

m-kuhn replied:
I guess you'll need a `QList<QgsExpressionNodeFunction> functionNodes( const QString &functionName = QString() /* if this is a null string return all functions */ )` or you'll get the function definition but no way to get to an evaluated parameter value.

elpaso committed:
Removed leftover

elpaso committed:
Removed const

I don't get why, but I trust @m-kuhn and I'll follow his advice :)

elpaso committed:
Moved form feature to private and added setters/getters

plus other minor style changes as suggested by m-kuhn
in his PR review

nyalldawson commented on resources/function_help/json/get_current_form_field_value.txt:
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "current_value",
+  "type": "function",
+  "description": "Returns the current value of a field in the form or table row currently being edited.",
```
I'd probably expand slightly on this for clarification:

"Returns the current, unsaved value of a field in the form or table row currently being edited. This will differ from the feature's actual attribute values for features which are currently being edited or have not yet been added to a layer."

nyalldawson commented on src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp:
```diff
@@ -162,3 +172,58 @@ QStringList QgsValueRelationFieldFormatter::valueToStringList( const QVariant &v
+
+QSet<QString> QgsValueRelationFieldFormatter::expressionFormVariables( const QString &expression )
+{
+  const QStringList formVariables( QgsExpressionContextUtils::formScope()->variableNames() );
```
Leaks a scope

nyalldawson commented on src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp:
```diff
@@ -162,3 +172,58 @@ QStringList QgsValueRelationFieldFormatter::valueToStringList( const QVariant &v
+  const QStringList formVariables( QgsExpressionContextUtils::formScope()->variableNames() );
+  QSet<QString> variables;
+
+  for ( auto const &variable : formVariables )
```
Instead of this, I'd go:
```
   std::unique_ptr< QgsExpressionContextScope > scope(QgsExpressionContextUtils::formScope() );
   QSet< QString > formVariables = scope->variableNames().toSet();
   const QSet< QString > usedVariables = QgsExpression( expression ).referencedVariables();
   formVariables.intersect( usedVariables );
   return formVariables;
```

nyalldawson commented on src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp:
```diff
@@ -162,3 +172,58 @@ QStringList QgsValueRelationFieldFormatter::valueToStringList( const QVariant &v
+QSet<QString> QgsValueRelationFieldFormatter::expressionFormAttributes( const QString &expression )
+{
+  QSet<QString> attributes;
+  const QStringList formFunctions( QgsExpressionContextUtils::formScope()->functionNames() );
```
leaks a scope

nyalldawson commented on src/core/fieldformatter/qgsvaluerelationfieldformatter.h:
```diff
@@ -62,23 +65,71 @@ class CORE_EXPORT QgsValueRelationFieldFormatter : public QgsFieldFormatter
+    /**
+     * Utility to convert an array or a string representation of and array \a value to a string list
+     *
+     * \param value The value to be converted
```
#pedantic, but there's no need to document params explicitly when you've already described them above

nyalldawson commented on src/core/fieldformatter/qgsvaluerelationfieldformatter.h:
```diff
@@ -62,23 +65,71 @@ class CORE_EXPORT QgsValueRelationFieldFormatter : public QgsFieldFormatter
+     * Utility to convert an array or a string representation of and array \a value to a string list
+     *
+     * \param value The value to be converted
+     * \return A string list
```
Drop this - the signature is already in the doc, so the return type is already visible

nyalldawson commented on src/core/fieldformatter/qgsvaluerelationfieldformatter.h:
```diff
@@ -62,23 +65,71 @@ class CORE_EXPORT QgsValueRelationFieldFormatter : public QgsFieldFormatter
+    /**
+     * Check whether the \a feature has all values required by the \a expression
+     *
+     * @return True if the expression can be used
```
\return

nyalldawson commented on src/core/fieldformatter/qgsvaluerelationfieldformatter.h:
```diff
@@ -62,23 +65,71 @@ class CORE_EXPORT QgsValueRelationFieldFormatter : public QgsFieldFormatter
+
+    /**
+     * Regular expression to find dynamic filtering based on form field values
+     * \see GetCurrentFormFieldValue()
```
GetCurrentFormFieldValue is private and hidden from the docs - so remove this line

nyalldawson commented on src/core/qgsexpressioncontext.cpp:
```diff
@@ -762,6 +789,16 @@ class GetProcessingParameterValue : public QgsScopedExpressionFunction
+QgsExpressionContextScope *QgsExpressionContextUtils::formScope( const QgsFeature &formFeature )
+{
+  QgsExpressionContextScope *scope = new QgsExpressionContextScope( QObject::tr( "Form" ) );
+  scope->setFeature( formFeature );
```
I think this can be neater:

    scope->setVariable( QStringLiteral("current_feature"), formFeature, true );

and then modify GetCurrentFormFieldValue to use 

   const QgsFeature feat( context->variable( QStringLiteral("current_feature") ) );

This avoids ANY chance that the feature will be overridden somewhere later in the context, and also means that a handy @current_feature variable will also be created

nyalldawson commented on src/gui/editorwidgets/core/qgseditorwidgetwrapper.h:
```diff
@@ -277,8 +278,44 @@ class GUI_EXPORT QgsEditorWidgetWrapper : public QgsWidgetWrapper
+    /**
+     * mFieldIdx the widget feature field id
+     */
+    int mFieldIdx;
```
initialize to -1

nyalldawson commented on tests/src/gui/testqgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -3,7 +3,9 @@
      --------------------------------------
     Date                 : 21 07 2017
     Copyright            : (C) 2017 Paul Blottiere
+    Copyright            : (C) 2018 Alessandro Pasotti
```
Do we do this? I always thought it was just whoever created the file gets added, and everyone else just accepts that :)

elpaso replied:
no problems to remove

elpaso committed:
Removed attribution

elpaso committed:
Add referencedFunctions() to expressions

elpaso committed:
Updated docs for current_value function

elpaso committed:
Added methods to search expression nodes and drop RegExp

- nodes() return the list of all nodes
- findNodes<T> returns a list of the nodes matching the class

Also drops the regexp for finding form attrs in the value-relation
expressions.

Other minor fixes as suggested in the PR review.

elpaso committed:
Initialize mFieldIdx to -1

elpaso committed:
Set form feature in current_feature variable instead of

... setting the feature member variable

nyalldawson commented on src/core/expression/qgsexpression.cpp:
```diff
@@ -771,6 +779,9 @@ void QgsExpression::initVariableHelp()
   sVariableHelpTexts.insert( QStringLiteral( "notification_message" ), QCoreApplication::translate( "notification_message", "Content of the notification message sent by the provider (available only for actions triggered by provider notifications)." ) );
+
+  //form context variable
+  sVariableHelpTexts.insert( QStringLiteral( "current_geometry" ), QCoreApplication::translate( "current_geometry", "Represents the geometry of the feature currently being edited in the form or the table row. Can be used for in a form/row context to filter the related features." ) );
```
You'll also need to add current_feature here

DelazJ replied:
"Can be used **for** in a form/row context to filter the related features."
Is that a correct sentence?

elpaso replied:
fixed in master

m-kuhn reviewed (approved):
Thanks for the efficient review process, looks very good to me now.

elpaso committed:
Added help text for @current_feature

elpaso commented:
Thanks to all contributors and to @m-kuhn and @nyalldawson for the review!

nyalldawson commented:
Thanks for your hard work here -- those extra expression functions will come in very handy!

phidrho commented on src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp:
```diff
@@ -213,19 +189,126 @@ void QgsValueRelationWidgetWrapper::setValue( const QVariant &value )
+}
+
+
+void QgsValueRelationWidgetWrapper::populate( )
```
This function is maybe related to #50997
